tree/dataframe/test/dataframe_snapshotWithVariations.cxx

```diff
@@ -0,0 +1,435 @@
+               ? entry
+               : ((systematicName.find("xVariation_0") != std::string::npos) ? entry + 1 : entry * 3);
+         const bool passCuts = (originalX % 2 == 0) || originalX == 5;
+
+         if (passCuts)
+            EXPECT_EQ(x, originalX) << "sys:'" << systematicName << "' originalX: " << originalX << " event: " << event;
+         else
+            EXPECT_EQ(x, 0) << "sys:'" << systematicName << "' originalX: " << originalX << " event: " << event;
+
+         ASSERT_EQ(y->size(), passCuts ? 4 : 0)
+            << "sys:'" << systematicName << "' entry: " << entry << " originalX: " << originalX << " event: " << event;
+         for (unsigned int i = 0; i < y->size(); ++i) {
+            EXPECT_EQ((*y)[i], x + i);
+         }
+      }
+      tree->ResetBranchAddresses();
+   }
+
+   file.reset();
+   gSystem->Unlink(filename);
+}
+
+TEST(RDFVarySnapshot, SnapshotVirtualClass)
+{
+   constexpr auto filename = "VarySnapshot_virtualClass.root";
+   ROOT::RDataFrame rdf(10);
+   ROOT::RDF::RSnapshotOptions opt;
+   opt.fIncludeVariations = true;
+
+   auto result =
+      rdf.Define("tf", [](ULong64_t entry) { return TwoFloats(entry); }, {"rdfentry_"})
+         .Vary("tf",
+               [](TwoFloats const &tf) {
+                  TwoFloats up, down;
+                  up.a = tf.a;
+                  up.b = tf.a + 1.f;
+                  down.a = tf.a;
+                  down.b = tf.a - 1.f;
+                  return ROOT::RVec<TwoFloats>{up, down};
+               },
+               {"tf"}, {"Up", "Down"})
+         .Filter([](TwoFloats const &tf) { return 2. * tf.a == tf.b || static_cast<unsigned int>(tf.a) % 2 == 0; },
+                 {"tf"})
+         .Snapshot("t", filename, {"tf", "rdfentry_"}, opt);
+
+   {
+      TFile file(filename, "READ");
+      std::unique_ptr<TTree> tree{file.Get<TTree>("t")};
+      ASSERT_TRUE(tree);
+      if (verbose) {
+         tree->Print("");
+         tree->Scan("*", "", "colsize=20");
+      }
+
+      TwoFloats *tf = nullptr, *tf_Up = nullptr, *tf_Down = nullptr;
+      ASSERT_EQ(tree->SetBranchAddress<TwoFloats>("tf", &tf), TTree::kMatch);
+      ASSERT_EQ(tree->SetBranchAddress<TwoFloats>("tf__tf_Up", &tf_Up), TTree::kMatch);
+      ASSERT_EQ(tree->SetBranchAddress<TwoFloats>("tf__tf_Down", &tf_Down), TTree::kMatch);
+      ASSERT_EQ(tree->GetEntries(), 10);
+      for (unsigned int i = 0; i < 10; ++i) {
+         tree->GetEntry(i);
+         EXPECT_EQ(tf->a, i);
+         EXPECT_EQ(tf->b, 2 * i);
+         if (i == 1) {
+            // Passes because a == 1 && b == 2 both for b=2*a and b=a+1
+            EXPECT_EQ(tf_Up->a, 1.f);
+            EXPECT_EQ(tf_Up->b, 2.f);
+         }
+         if (i % 2 == 0) {
+            EXPECT_EQ(tf_Up->a, i);
+            EXPECT_EQ(tf_Up->b, i + 1.f);
+            EXPECT_EQ(tf_Down->a, i);
+            EXPECT_EQ(tf_Down->b, i - 1.f);
+         }
+      }
+   }
+
+   if (!HasFailure())
+      gSystem->Unlink(filename);
```

Comment:
Here and in all similar `Unlink`/`std::remove` calls, I would propose to have a RAII structure to manage file cleanup to be more exception-safe